File: Comandas/Api/dotnet_api/src/Models/Caja.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace BarIceCreamShop.Api.Models;

[Table("cajas")]
public class Caja
{
    [Column("id_caja")] public int Id { get; set; }
    [Column("id_empresa")] public int IdEmpresa { get; set; }
    [Column("id_sucursal")] public int IdSucursal { get; set; }
    [Column("estado")] public string Estado { get; set; } = "Abierta";
    [Column("id_usuario_apertura")] public int UsuarioAperturaId { get; set; }
    [Column("fecha_apertura")] public DateTime FechaApertura { get; set; }
    [Column("monto_inicial", TypeName = "decimal(18,2)")] public decimal MontoInicial { get; set; }
    [Column("id_usuario_cierre")] public int? UsuarioCierreId { get; set; }
    [Column("fecha_cierre")] public DateTime? FechaCierre { get; set; }
    [Column("total_ventas", TypeName = "decimal(18,2)")] public decimal? TotalVentas { get; set; }
    [Column("total_ingresos", TypeName = "decimal(18,2)")] public decimal? TotalIngresos { get; set; }
    [Column("total_egresos", TypeName = "decimal(18,2)")] public decimal? TotalEgresos { get; set; }
    [Column("total_efectivo_esperado", TypeName = "decimal(18,2)")] public decimal? TotalEfectivoEsperado { get; set; }
    [Column("observaciones")] public string? Observaciones { get; set; }

    public Usuario UsuarioApertura { get; set; } = null!;
    public Usuario? UsuarioCierre { get; set; }
}

[Table("caja_movimientos")]
public class CajaMovimiento
{
    [Column("id_movimiento")] public int Id { get; set; }
    [Column("id_caja")] public int CajaId { get; set; }
    [Column("tipo")] public string Tipo { get; set; } = "Ingreso";
    [Column("monto", TypeName = "decimal(18,2)")] public decimal Monto { get; set; }
    [Column("motivo")] public string Motivo { get; set; } = string.Empty;
    [Column("id_usuario")] public int UsuarioId { get; set; }
    [Column("fecha")] public DateTime Fecha { get; set; }

    public Usuario Usuario { get; set; } = null!;
}

[Table("caja_cierre_detalle")]
public class CajaCierreDetalle
{
    [Column("id_caja")] public int CajaId { get; set; }
    [Column("medio_pago")] public string MedioPago { get; set; } = string.Empty;
    [Column("total", TypeName = "decimal(18,2)")] public decimal Total { get; set; }
}

File: Comandas/App/bar-icecream-shop/flutter_app/lib/models/caja.dart
class CajaMedioPago {
  final String medioPago;
  final double total;

  const CajaMedioPago({required this.medioPago, required this.total});

  factory CajaMedioPago.fromJson(Map<String, dynamic> json) => CajaMedioPago(
        medioPago: json['medioPago']?.toString() ?? '',
        total: (json['total'] as num?)?.toDouble() ?? 0,
      );
}

class CajaMovimiento {
  final int id;
  final String tipo;
  final double monto;
  final String motivo;
  final String usuario;
  final DateTime fecha;

  const CajaMovimiento({
    required this.id,
    required this.tipo,
    required this.monto,
    required this.motivo,
    required this.usuario,
    required this.fecha,
  });

  bool get isIngreso => tipo == 'Ingreso';

  factory CajaMovimiento.fromJson(Map<String, dynamic> json) => CajaMovimiento(
        id: (json['id'] as num?)?.toInt() ?? 0,
        tipo: json['tipo']?.toString() ?? '',
        monto: (json['monto'] as num?)?.toDouble() ?? 0,
        motivo: json['motivo']?.toString() ?? '',
        usuario: json['usuario']?.toString() ?? '',
        fecha: DateTime.parse(json['fecha'].toString()).toLocal(),
      );
}

class Caja {
  final int id;
  final String estado;
  final String usuarioApertura;
  final DateTime fechaApertura;
  final double montoInicial;
  final String? usuarioCierre;
  final DateTime? fechaCierre;
  final double? totalVentas;
  final double? totalIngresos;
  final double? totalEgresos;
  final double? totalEfectivoEsperado;
  final String? observaciones;
  final List<CajaMedioPago> detalle;
  final List<CajaMovimiento> movimientos;

  const Caja({
    required this.id,
    required this.estado,
    required this.usuarioApertura,
    required this.fechaApertura,
    required this.montoInicial,
    this.usuarioCierre,
    this.fechaCierre,
    this.totalVentas,
    this.totalIngresos,
    this.totalEgresos,
    this.totalEfectivoEsperado,
    this.observaciones,
    this.detalle = const [],
    this.movimientos = const [],
  });

  bool get abierta => estado == 'Abierta';

  factory Caja.fromJson(Map<String, dynamic> json) => Caja(
        id: (json['id'] as num?)?.toInt() ?? 0,
        estado: json['estado']?.toString() ?? 'Abierta',
        usuarioApertura: json['usuarioApertura']?.toString() ?? '',
        fechaApertura:
            DateTime.parse(json['fechaApertura'].toString()).toLocal(),
        montoInicial: (json['montoInicial'] as num?)?.toDouble() ?? 0,
        usuarioCierre: json['usuarioCierre']?.toString(),
        fechaCierre: json['fechaCierre'] == null
            ? null
            : DateTime.parse(json['fechaCierre'].toString()).toLocal(),
        totalVentas: (json['totalVentas'] as num?)?.toDouble(),
        totalIngresos: (json['totalIngresos'] as num?)?.toDouble(),
        totalEgresos: (json['totalEgresos'] as num?)?.toDouble(),
        totalEfectivoEsperado:
            (json['totalEfectivoEsperado'] as num?)?.toDouble(),
        observaciones: json['observaciones']?.toString(),
        detalle: (json['detalle'] as List<dynamic>? ?? const [])
            .map((e) => CajaMedioPago.fromJson(e as Map<String, dynamic>))
            .toList(),
        movimientos: (json['movimientos'] as List<dynamic>? ?? const [])
            .map((e) => CajaMovimiento.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
